lib/tools/system-calls.js

```diff
@@ -477,7 +477,18 @@ systemCallMethods.waitForDevice = async function (appDeviceReadyTimeout = 30) {
 };
 
 systemCallMethods.reboot = async function () {
-  await this.shell(['stop']);
+  try {
+    await this.shell(['stop']);
+  } catch (err) {
+    if (err.message.indexOf('must be root') === -1) {
+      throw err;
+    }
+    // this device needs adb to be running as root to stop.
+    // so try to restart the daemon
+    log.debug('Device requires adb to be running as root in order to reboot. Restarting daemon');
+    await exec(this.executable.path, ['root']);
```

Comment:
@imurchie should we run `await exec(this.executable.path, ['unroot']);` after the device reboots to avoid mean people trying to do bad things with root privileges?